### ISSUE TITLE
Neighbourhood and name based location search on Map Page

### ISSUE THREAD

**Issue by gkjohn** [klpdotorg/dubdubdub#233]

**Comment by batpad** [comment]
So. "Integrating" neighbourhood / name based location search using the Google geocoder and our own entity search through the KLP db is a bit tricky.

Problem being that it is not possible for us to do a geocode query to google from the back-end every time an autocomplete request is made. This would both slow down the API, but also, practically, hit Google usage limits very quickly (if we are trying to autocomplete using server-side geocoding).

To do the geocoding on the client side, makes it kinda impossible to seamlessly integrate with the existing select2 widget and search (querying two different end-points and populating the same autocomplete widget just seems a bit improbable / impossible)

Plus, the user experience when you click on a google returned location and when you click on it is a bit different from clicking on a KLP entity (clicking on a KLP entity will go to that feature on the map and generally open a popup on that location, clicking on a google location result will just center on that point).

One idea was to only trigger the location search if our API returns no results - either through explicit user interaction, or automatically, and then have that behave a bit differently (bit of a design question for @meghaklp ).

Another option would be to have a different search box for this search, although I agree that this is totally weird from a user perspective - not sure if there are any creative design solutions to that.

Related question: by Google's TOS, do we need to show results on a Google map / show google branding anywhere? @geohacker 

/cc @gkjohn @meghaklp

**Comment by gkjohn** [comment]
@geohacker Does OSM have place/area/locality/neighbourhood search? No matter how good or bad it is? @batpad, if OSM does, can we use that? And yes, while including it in the existing search box results is confusing, it's better than not having it at all. And when clicked, map will pan to that area and populate pins relevant to that zoom level. 

![screenshot 2015-07-30 07 53 53](https://cloud.githubusercontent.com/assets/1021034/8974530/3dffe8cc-3690-11e5-81a9-7935b05b5e4f.png)

When we do the Advanced Search with @meghaklp, we can improve this.

**Comment by batpad** [comment]
@geohacker is it possible / feasible to use some OSM-based free geocoding to do that server-side and add it to the dictionary of results the API sends back? Please advise :). Thank you.

**Comment by gkjohn** [comment]
@meghaklp Here's a thought. What if we had a toggle in the search box that you can toggle between Search Schools and Search Localities. The former is what we currently have. The latter is a toggle to search based on areas etc. @batpad 

![screenshot 2015-08-18 12 16 47](https://cloud.githubusercontent.com/assets/1021034/9323926/2bb11aa2-45a3-11e5-8da2-38fc8132524e.png)

**Comment by meghaklp** [comment]
Toggle - meaning like a select box or a radio (if it's a limited number of options) . Can do just above the search box, like this?
![search](https://cloud.githubusercontent.com/assets/465382/9330654/60e8d43c-45d8-11e5-95f1-3e123f556081.png)

**Comment by gkjohn** [comment]
@meghaklp Yep! But doing this depends on @batpad figuring out how to do the locality search.
